### ISSUE TITLE
Grakn 2.0 LOCK; [SRV03] Invalid Server Operation while loading data into grakn 2.0

### ISSUE THREAD

**Issue by gunnarklee** [graknlabs/grakn#6006]
## Description

While loading data into grakn 2.0 the server locked and I am unable to get it to start up again. replacing the grakn-core-all file does not fix the problem since the database seems to still be intact

## Environment

1. OS Ubuntu 16.4
2. Grakn version (and platform) e.g. Grakn Core 2.0.0-alpha (1/9/21)
3. Grakn client: client-python 2.0.0.4a
4. Other environment details: console runs

## Reproducible Steps
1. bash ~/grakn-core/grakn-core-all-linux-2.0.0-alpha/grakn server

Steps to create the smallest reproducible scenario:
1. bash ~/grakn-core/grakn-core-all-linux-2.0.0-alpha/grakn serve

## Expected Output
expected server to run

## Actual Output
**before replacing the grakn-core-all**

```
====================================================================================================
      ________  _____     _______  __    __  __    __      _______  _______  _____     _______
     |   __   ||   _  \  |   _   ||  |  /  /|  \  |  |    |   _   ||   _   ||   _  \  |   ____|
     |  |  |__||  | |  | |  | |  ||  | /  / |   \ |  |    |  | |__||  | |  ||  | |  | |  |
     |  | ____ |  |_| /  |  |_|  ||  |/  /  |        |    |  |     |  | |  ||  |_| /  |  |____
     |  ||_   ||   _  \  |   _   ||   _  \  |   _    |    |  |  __ |  | |  ||   _  \  |   ____|
     |  |__|  ||  | \  \ |  | |  ||  | \  \ |  | \   |    |  |_|  ||  |_|  ||  | \  \ |  |____
     |________||__|  \__\|__| |__||__|  \__\|__|  \__|    |_______||_______||__|  \__\|_______|

                                         THE KNOWLEDGE GRAPH
====================================================================================================
org.rocksdb.RocksDBException: While lock file: /home/ubuntu/grakn-core/grakn-core-all-linux-2.0.0-alpha/server/data/rheos_v3/schema/LOCK: Resource temporarily unavailable
[SRV03] Invalid Server Operation: Exited with error.

```

**after replacing the grakn-core-all**

```
====================================================================================================
      ________  _____     _______  __    __  __    __      _______  _______  _____     _______
     |   __   ||   _  \  |   _   ||  |  /  /|  \  |  |    |   _   ||   _   ||   _  \  |   ____|
     |  |  |__||  | |  | |  | |  ||  | /  / |   \ |  |    |  | |__||  | |  ||  | |  | |  |
     |  | ____ |  |_| /  |  |_|  ||  |/  /  |        |    |  |     |  | |  ||  |_| /  |  |____
     |  ||_   ||   _  \  |   _   ||   _  \  |   _    |    |  |  __ |  | |  ||   _  \  |   ____|
     |  |__|  ||  | \  \ |  | |  ||  | \  \ |  | \   |    |  |_|  ||  |_|  ||  | \  \ |  |____
     |________||__|  \__\|__| |__||__|  \__\|__|  \__|    |_______||_______||__|  \__\|_______|

                                         THE KNOWLEDGE GRAPH
====================================================================================================
Failed to bind
java.io.IOException: Failed to bind
	at io.grpc.netty.NettyServer.start(NettyServer.java:264)
	at io.grpc.internal.ServerImpl.start(ServerImpl.java:183)
	at io.grpc.internal.ServerImpl.start(ServerImpl.java:90)
	at grakn.core.server.GraknServer.start(GraknServer.java:284)
	at grakn.core.server.GraknServer.startGraknServer(GraknServer.java:243)
	at grakn.core.server.GraknServer.main(GraknServer.java:205)
Caused by: java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:461)
	at sun.nio.ch.Net.bind(Net.java:453)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:222)
	at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:550)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:248)
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)
Failed to bind
[SRV03] Invalid Server Operation: Exited with error.

Shutting down Grakn Core Server...
Grakn Core Server has been shutdown
 
```

## Additional information

since the console works and i see the database that must be locked i tried to delete the database but could not get to it.

in the console: 
> help

database list                              List the databases on the server
database create <db>                       Create a database with name <db> on the server
database delete <db>                       Delete a database with name <db> on the server
transaction <db> schema|data read|write    Start a transaction to database <db> with schema or data session, with read or write transaction
help                                       Print this help menu
clear                                      Clear console screen
exit                                       Exit console

> database list
io exception

> database delete keyspacename_v3
io exception

**Comment by gunnarklee** [comment]
after about 10 or 15 min the server became responsive again but my keyspace was gone 

>database list 
>

**Comment by gunnarklee** [comment]
this happened a second time while trying to load data to the grakn V2.0 server. I am using multi threading but keeping batch size below 200. 

ON THE SERVER
org.rocksdb.RocksDBException: While lock file: /home/ubuntu/grakn-core/grakn-core-all-linux-2.0.0-alpha/server/data/XXXXXXX/schema/LOCK: Resource temporarily unavailable
[SRV03] Invalid Server Operation: Exited with error.

ON THE CLIENT SHELL
>>> ncbi.load_ncbi_gene(conn, client, keyspace)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/grakn_pilot/loaders/load_ncbi.py", line 64, in load_ncbi_gene
    gu.batch_load_queries(client, keyspace, batches_iterator)
  File "/home/ubuntu/grakn_pilot/loaders/grakn_utils.py", line 317, in batch_load_queries
    with client.session(keyspace, SessionType.DATA) as session:
  File "/home/ubuntu/anaconda3/envs/grakn2/lib/python3.6/site-packages/grakn/client.py", line 41, in session
    return _Session(self, database, session_type, options)
  File "/home/ubuntu/anaconda3/envs/grakn2/lib/python3.6/site-packages/grakn/rpc/session.py", line 53, in __init__
    self._session_id = self._grpc_stub.session_open(open_req).session_id
  File "/home/ubuntu/anaconda3/envs/grakn2/lib/python3.6/site-packages/grpc/_channel.py", line 923, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/home/ubuntu/anaconda3/envs/grakn2/lib/python3.6/site-packages/grpc/_channel.py", line 826, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNAVAILABLE
        details = "failed to connect to all addresses"
        debug_error_string = "{"created":"@1610314399.407584461","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":4165,"referenced_errors":[{"created":"@1610314399.407578482","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":397,"grpc_status":14}]}"
        
FINALLY 
and as before the server seems to have unlocked itself after 10 min and the databases were intact this time

**Comment by lolski** [comment]
> While loading data into grakn 2.0 the server locked and I am unable to get it to start up again

Could you elaborate on what exactly you mean when you said that "the server locked"? Did it crashed? Or did it stall? Is there an error message that was printed?

> replacing the grakn-core-all file does not fix the problem since the database seems to still be intact

Could you elaborate on what did you mean by "replacing the grakn-core-all file? Did you delete the whole directory and re-unzip a new one?

**Comment by gunnarklee** [comment]
> > While loading data into grakn 2.0 the server locked and I am unable to get it to start up again
> 
> Could you elaborate on what exactly you mean when you said that "the server locked"? Did it crashed? Or did it stall? Is there an error message that was printed?

on the client it is the error message shown above but I think the pertinent part is 

grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
status = StatusCode.UNAVAILABLE

on the console window, I see the SRV03 error both times.

org.rocksdb.RocksDBException: While lock file: /home/ubuntu/grakn-core/grakn-core-all-linux-2.0.0-alpha/server/data/XXXXXXX/schema/LOCK: Resource temporarily unavailable
[SRV03] Invalid Server Operation: Exited with error

If I kill the server window then restart it I get the same "[SRV03] Invalid Server Operation: Exited with error"

I think this is stalling with a lock. After 10-15 min it is able to be started.


> 
> > replacing the grakn-core-all file does not fix the problem since the database seems to still be intact
> 
> Could you elaborate on what did you mean by "replacing the grakn-core-all file? Did you delete the whole directory and create a new one by unzipping the zip / tar.gz distribution?

I unzipped a fresh copy of the whole directory and changed the name of the old directory. I think this is irrelevant because I go the same restart when i left the whole folder intact later.

**Comment by gunnarklee** [comment]
James mentioned that this happens with an open schema. I don't think this is related to opening the schema since I was loading data and it was working until it was not. More like a memory overload.

**Comment by lolski** [comment]
Based on a call with Gunnar, we've clarified the issues we've had above, in which the following conclusion was made:

- [Grakn hangs] Grakn works most of the time but sometimes it does hang. We'll have to monitor when this issue pops up again.
- [Deleted data] The issue where the data seems to get deleted is most probably because the distribution folder was moved. The data is located within the distribution and therefore it is also moved to the new location.
- [How to kill an unresponsive Grakn] We need to make sure to kill Grakn properly (using the kill -9 <grakn-process-id> command) when it's unresponsive. We can verify if it's properly terminated with ps -ef | grep GraknServer afterwards.
- [How to release schema session] We need to make sure that the application closes any schema session properly. If unable to do so, restart Grakn to release the schema session
